src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/BotSegmentZipParser.cs

```diff
@@ -30,24 +85,61 @@ public static List<BotSegment> ParseBotSegmentZipFromSystemPath(string zipFilePa
                 {
                     using var sr = new StreamReader(entry.Open());
-                    var frameData = JsonConvert.DeserializeObject<BotSegment>(sr.ReadToEnd());
 
-                    if (frameData.EffectiveApiVersion > SdkApiVersion.CURRENT_VERSION)
+                    var fileText = sr.ReadToEnd();
+
+                    try
                     {
-                        versionMismatch = frameData.EffectiveApiVersion;
-                        break;
-                    }
+                        // see if this is a bot segment list file
+                        var botSegmentList = JsonConvert.DeserializeObject<BotSegmentList>(fileText);
 
-                    badSegmentName = entry.FullName;
+                        if (botSegmentList.EffectiveApiVersion > SdkApiVersion.CURRENT_VERSION)
+                        {
+                            versionMismatch = botSegmentList.EffectiveApiVersion;
+                            break;
+                        }
 
-                    frameData.Replay_SegmentNumber = replayNumber++;
+                        foreach (var botSegment in botSegmentList.segments)
+                        {
+                            botSegment.Replay_SegmentNumber = replayNumber++;
 
-                    if (sessionId == null)
+                            if (sessionId == null)
+                            {
+                                sessionId = botSegment.sessionId;
+                            }
+
+                            results.Add(botSegment);
+                        }
+                    }
+                    catch (Exception)
                     {
-                        sessionId = frameData.sessionId;
+                        try
+                        {
+                            // not a bot segment list, must be a regular segment
+                            var frameData = JsonConvert.DeserializeObject<BotSegment>(fileText);
+
+                            if (frameData.EffectiveApiVersion > SdkApiVersion.CURRENT_VERSION)
+                            {
+                                versionMismatch = frameData.EffectiveApiVersion;
+                                break;
+                            }
+
+                            frameData.Replay_SegmentNumber = replayNumber++;
+
+                            if (sessionId == null)
+                            {
+                                sessionId = frameData.sessionId;
+                            }
+
+                            results.Add(frameData);
+                        }
+                        catch (Exception ex)
+                        {
+                            throw new Exception($"Invalid/missing file data while parsing BotSegment or BotSegmentList from zip file path: {zipFilePath} at entry: {entry.Name}", ex);
+                        }
                     }
 
-                    results.Add(frameData);
+                    badSegmentName = entry.FullName;
```

Comment:
If a version mismatch error happens above, then the loop would break before `badSegmentName` gets set. Perhaps this should be happening at the top of the loop?

Reply:
yep, that's from before i made this throw.. will fix, good catch